### PR TITLE
Add docker features (resolves #142, resolves #173, resolves #190)

### LINKED ISSUE #142: Build dockerized protect for every release

### LINKED ISSUE #173: Automatically generate docker image for major releases

### LINKED ISSUE #190: Add fusion options to dockerfile
continues #92

### PR DESCRIPTION
**Pull request by aofarrel**

### CODE REVIEW

**Comment by arkal** [comment]
Jenkins, test this please.

File: docker/wrapper.py [hunk 1]
```diff
@@ -82,6 +82,12 @@
             version: {samtools_somaticsniper_ver}
         bam_readcount:
             version: {bamreadcount_ver}
+    star_fusion:
+        run: {starfusion}
+        version: {star_fusion_ver}
+    fusion_inspector:
+        run: {fusioninspector}
```
Comment (arkal) [inline docker/wrapper.py]: `run_trinity`

**Comment by aofarrel** [comment]
@arkal This is ready.

File: docker/pipelineWrapper.py [hunk 2]
```diff
@@ -218,3 +225,30 @@ def check_for_input(tool_input, name):
...
+        and allows properly formatted descriptions ot be printed without line
+        wrapping.
+        '''
+        help = action.help
+        if '%(default)' not in action.help:
+            if action.default is not argparse.SUPPRESS:
+                defaulting_nargs = [argparse.OPTIONAL, argparse.ZERO_OR_MORE]
+                if action.option_strings or action.nargs in defaulting_nargs:
+                    help += ' (default: %(default)s)'
+        return help
```
Comment (arkal) [inline docker/pipelineWrapper.py]: add newline

File: docker/protect.cwl [hunk 3]
```diff
@@ -152,23 +269,77 @@ inputs:
...
+      prefix: --run-trinity
+
   snpeff:
     type: File?
     doc: "snpeff"
     inputBinding:
       prefix: --snpeff
 
-  transgene:
+  transgene-pep:
```
Comment (arkal) [inline docker/protect.cwl]: use the whole name. transgene-peptide-fasta

File: docker/protect.cwl [hunk 4]
```diff
@@ -152,23 +269,77 @@ inputs:
...
     type: File?
     doc: "transgene"
     inputBinding:
-      prefix: --transgene
+      prefix: --transgene-pep
+  transgene_trans:
```
Comment (arkal) [inline docker/protect.cwl]: see above

File: docker/protect.cwl [hunk 5]
```diff
@@ -152,23 +269,77 @@ inputs:
...
+  type: File?
+    inputBinding:
+      --transgene-trans
+  transgene_anno:
```
Comment (arkal) [inline docker/protect.cwl]: see above

File: docker/wrapper.py [hunk 6]
```diff
@@ -1,136 +1,433 @@
...
-    java_Xmx: 20G
+    dockerhub: {dockerhub}
+    java_Xmx: {javaxmx}
+    reference_build: {reference_build}
     sse_key: {sse_key}
     sse_key_is_master: {sse_key_is_master}
-    cghub_key:
-    storage_location: Local
+    gdc_download_token: {gdc_download_token}
+    storage_location: {storage_location}
```
Comment (arkal) [inline docker/wrapper.py]: revert this back to Local.

File: docker/wrapper.py [hunk 7]
```diff
@@ -1,136 +1,433 @@
...
+            tndelta: {mhcii_tndelta}
+
+reports:
+    mhc_pathways_file: {mhc_pathways_file}
+    itx_resistance_file: {itx_resistance_file}
+    immune_resistance_pathways_file: {immune_resistance_pathways_file}
+    car_t_targets_file: {car_t_targets_file}""")
+
+
+def str2bool(v):
```
Comment (arkal) [inline docker/wrapper.py]: Add a docstring, add the stackoverflow link to the question where you found this answer to the docstring

File: docker/wrapper.py [hunk 8]
```diff
@@ -1,136 +1,433 @@
...
+                        help='')
+
+    parser.add_argument('--dockerhub', type=str, default='aarjunrao')
+    parser.add_argument('--javaxmx', default='20G', type=str)
+    parser.add_argument('--sse-key', type=str, default='', help='Path to the desired SSE-key, if any.')
+    parser.add_argument('--sse-key-is-master', type=str2bool, default='False',
                         help='Indicates if the passed sse-key is the master key.')
-    parser.add_argument('--sse-key', type=str, default='',
-                        help='Path to the desired SSE-key, if any.')
+    parser.add_argument('--gdc-download-token', type=str)
```
Comment (arkal) [inline docker/wrapper.py]: help='A download token used to download files from the GDC'

File: docker/wrapper.py [hunk 9]
```diff
@@ -1,136 +1,433 @@
...
+    parser.add_argument('--storage-location', type=str, default="Local",
+                        help='Local or aws:<bucket_name> for where the output must go')
 
     parser.add_argument('--work-mount', required=True,
                         help='Mount where intermediate files should be written. This directory '
                              'should be mirror mounted into the container.')
 
     args = parser.parse_args()
+    #print(args)
```
Comment (arkal) [inline docker/wrapper.py]: remove this

File: docker/wrapper.py [hunk 10]
```diff
@@ -1,136 +1,433 @@
...
-    mhci_combo : 0.0,0.0,0.32,0.0,0.68
-    mhcii_combo : 0.25,0.25,0.25,0.25
-
-mhc_pathway_assessment:
-    genes_file : {mhc_pathway_assessment}""")
 
 if __name__ == '__main__':
+    import protect
+    with open(os.path.join(os.path.dirname(protect.__file__), "pipeline", "defaults.yaml")) as def_file:
+        #print(os.path.join(os.path.dirname(protect.__file__), "pipeline"))
```
Comment (arkal) [inline docker/wrapper.py]: remove all these print statements.

**Comment by arkal** [comment]
Jenkins, test this please.

**Comment by arkal** [comment]
Jenkins, test this please.

File: docker/wrapper.py (hunk 7 above)
Comment (arkal) [inline docker/wrapper.py]: resolved

**Pull request merged**